aggregator/follower_flush_mgr.go

```diff
@@ -0,0 +1,344 @@
+	}
+	mgr.flushTask.flushersByInterval = flushersByInterval
+	return mgr.flushTask, 0
+}
+
+// NB(xichen): the follower flush manager flushes data based on the flush times
+// stored in kv and does not need to take extra actions when a new bucket is added.
+func (mgr *followerFlushManager) OnBucketAdded(int, *flushBucket) {}
+
+// The follower flush manager may only lead if and only if all the following conditions
+// are met:
+// * The flush times persisted in kv have been read at least once.
+// * There are no outstanding/unprocessed kv flush times update.
+// * All the aggregation windows since the flush manager is opened have ended.
+func (mgr *followerFlushManager) CanLead() bool {
+	mgr.RLock()
+	defer mgr.RUnlock()
+
+	if mgr.flushTimesState != flushTimesProcessed {
+		mgr.metrics.flushTimesNotProcessed.Inc(1)
+		return false
+	}
+
+	allWindowsEnded := true
+	for _, shardFlushTimes := range mgr.proto.ByShard {
+		for windowNanos, lastFlushedNanos := range shardFlushTimes.ByResolution {
+			windowSize := time.Duration(windowNanos)
+			windowEndAt := mgr.openedAt.Truncate(windowSize)
+			if windowEndAt.Before(mgr.openedAt) {
+				windowEndAt = windowEndAt.Add(windowSize)
+			}
+			if lastFlushedNanos < windowEndAt.UnixNano() {
+				mgr.metrics.flushWindowsNotEnded.Inc(1)
+				allWindowsEnded = false
+				break
+			}
+		}
+		if !allWindowsEnded {
+			break
+		}
+	}
+	return allWindowsEnded
+}
+
+func (mgr *followerFlushManager) flushersFromKVUpdateWithLock(buckets []*flushBucket) []flushersGroup {
+	flushersByInterval := make([]flushersGroup, len(buckets))
+	for i, bucket := range buckets {
+		flushersByInterval[i].interval = bucket.interval
+		flushersByInterval[i].duration = bucket.duration
+		flushersByInterval[i].flushers = make([]flusherWithTime, 0, defaultInitialFlushTimesCapacity)
+		for _, flusher := range bucket.flushers {
+			shard := flusher.Shard()
+			shardFlushTimes, exists := mgr.proto.ByShard[shard]
+			if !exists {
+				mgr.metrics.shardNotFound.Inc(1)
+				mgr.logger.WithFields(
+					xlog.NewLogField("shard", shard),
+				).Warn("shard not found in flush times")
+				continue
+			}
+			resolution := flusher.Resolution()
+			lastFlushedAtNanos, exists := shardFlushTimes.ByResolution[int64(resolution)]
+			if !exists {
+				mgr.metrics.resolutionNotFound.Inc(1)
+				mgr.logger.WithFields(
+					xlog.NewLogField("shard", shard),
+					xlog.NewLogField("resolution", resolution.String()),
+				).Warn("resolution not found in flush times")
+				continue
+			}
+			newFlushTarget := flusherWithTime{
+				flusher:          flusher,
+				flushBeforeNanos: lastFlushedAtNanos,
+			}
+			flushersByInterval[i].flushers = append(flushersByInterval[i].flushers, newFlushTarget)
+		}
+	}
+	return flushersByInterval
+}
+
+func (mgr *followerFlushManager) flushersFromForcedFlushWithLock(
+	buckets []*flushBucket,
+	flushBeforeNanos int64,
+) []flushersGroup {
+	flushersByInterval := make([]flushersGroup, len(buckets))
```

Comment:
Why does this need to be done with a lock? It looks like this func is generating flushersByInterval using params, I don't see any reads of anything off mgr.